Patch 1

papofrito opened the pull request:
I'm fan of grid pages, so I thought it would be lovely "Figure out how to find approriate remote parent id" :) Hope it helps.

papofrito committed:
Update WPCrosspostAPIClient.php

papofrito committed:
Update wp-crosspost.php

meitar commented:
Thanks for your contribution! :) I think this would be a nice addition to the project, but I have some concerns about the implementation you've proposed here.

It looks like you're defining a method that talks to the remote host every to retrieve metadata every time WP Crosspost prepares a post to be crossposted. This means that a single cross-post now requires two HTTP round trips, instead of one. This seems unnecessary or at least like it would put unnecessary load on the remote server and significantly slow crossposting down. Also, what if we want to get more metadata? The amount of HTTP round-trips we'd need increases linearly for each new piece of metadata we fetch, and that's *before* crossposting.

Instead of fetching this information every time for each cross-post, have you considered saving the result somewhere in the post metadata itself so it doesn't need to be refetched each time, and then updating it somehow?

Have you considered any of the above? If so, can you tell me more about why you chose to go with this implementation?

Thanks again!

papofrito commented:
Of course, you are right, probably that's why I'm not a programmer :p I was looking for a minimal aproach. I'll take a look to the wp metadata api to implement your welcomed suggestion :)

meitar commented:
Great. :) Have a look through the various [`*post_meta` functions and hooks](https://developer.wordpress.org/?s=post_meta&post_type[]=wp-parser-function&post_type[]=wp-parser-hook) in the WordPress developer reference to see what you can do with WP's metadata. Let me know if you run into problems or have questions.

papofrito committed:
Update WPCrosspostAPIClient.php

I realized that there is no need for an additional function to get remote parent page id by slug

papofrito committed:
Update wp-crosspost.php

Get remote parent post id from local post metadata => only set remote parent if local parent is previously crossposted

papofrito commented:
First approach was very wrong, changed to get remote parent from local post parent metadata. No need for any new function, only one line of code in wp-crosspost.php. parent post for the entry is only send to wp.com if parent post was previously crossposted, there is no additional remote request :)

meitar commented:
This is definitely better! I think there are still problems, though.

My understanding of this change is that it will:

1. Look at the current post's `post_parent` field to see if the post itself has a parent specified.
1. If the current post does have a parent (i.e., this post is hierarchically situated), it asks the parent post for its remote post ID.
 1. If the parent post has a remote post ID, the current post's `parent` is set to that remote post when crossposting.
1. If the current post does *not* have a parent, the current post's `parent` is set to the empty string, which means "no parent."

This seems like it would work well for one-directional cross-posting, but have you tested this going the other way? I.e., from the remote site back into the local site?

In other words, what happens to the current post hierarchy when WP Crosspost imports the remote post into the current site? I'm worried (though I have not run the code to verify) that:

1. When the `importPostFromWordPress()` method is run, the remote post's `post_parent` ID is set as the local post's `post_parent`.
1. When the local post is edited, it will overwrite the data in the `parent` field in the API call to the remote blog to the post ID of the *local* parent, rather than the remote parent ID.

To be honest I haven't taken a close enough look at the code I wrote in `importPostFromWordPress()` to know if I got this right, and I haven't done any testing on parent posts, so it's possible that your change is all that's needed. But I want to be certain that this change works for all functionality (creating a new post, updating/editing an existing post, importing a remote post, etc.) before merging.

Have you already run or can you show me a test of the above? :)

Thank you again for your work on this feature!
